include/libkrun.h

```diff
@@ -113,6 +120,41 @@ int32_t krun_set_data_disk(uint32_t ctx_id, const char *disk_path);
  */
 int32_t krun_add_disk(uint32_t ctx_id, const char *block_id, const char *disk_path, bool read_only);
 
+/* Supported disk image formats */
+#define KRUN_DISK_FORMAT_UNSPECIFIED 0
+#define KRUN_DISK_FORMAT_RAW 1
+#define KRUN_DISK_FORMAT_QCOW2 2
+/**
+ * Adds a disk image to be used as a general partition for the microVM. The supported
+ * image formats are: "raw" and "qcow2".
+ *
+ * This API is mutually exclusive with the deprecated krun_set_root_disk and
+ * krun_set_data_disk methods and must not be used together.
+ *
+ * SECURITY NOTE:
+ * By all possible means, one should avoid probing the disk image that will be given to libkrun.
+ * If a raw file is probed, and a QCOW2 header is written into the raw file's header by the guest,
+ * the next time libkrun launches the VM it will boot as a QCOW2 image. QCOW2 files can, and will,
+ * open other files. The guest will most likely have access to the data in those files, essentially
+ * giving the guest access to any file on the host system the disk emulation has access to.
+ *
+ * Arguments:
+ *  "ctx_id"      - the configuration context ID.
+ *  "block_id"    - a null-terminated string representing the partition.
+ *  "disk_path"   - a null-terminated string representing the path leading to the disk image.
+ *  "disk_format" - the disk image format
+ *  "read_only"   - whether the mount should be read-only. Required if the caller does not have
+ *                  write permissions (for disk images in /usr/share).
+ *
+ * Returns:
+ *  Zero on success or a negative error number on failure.
+ */
+int32_t krun_add_disk2(uint32_t ctx_id,
```

Comment:
Although not very nice and desperately needing a change when revamp the API for libkrun 2.0, we *could* just check the file extension for `.qcow2` or `.raw` instead of adding another API.

Reply:
I don't think that is a good idea, it's like probing but worse. Also, extensions means nothing in a unix world, for instance, I have many qcow2 and raw images sharing the same extension `.img` 


>> @mtjhrc wrote: I am not sure if libkrun should provide the ability to probe the image.

> @jakecorrenti wrote: To me this feels like we are putting a lot of trust on the user to do the right thing.

What I suggested, is to spilt the API into probing and adding so you will have something like:
```
int32_t krun_add_disk2(..., format format_type)  // Add a disk with a specified format
format_type krun_detect_format(...)  // Returns the format
```
(maybe detect_format should be part of imago?)

So, the user has the flexibility to do probing or strictly only accepting a disk + its format

Sadly this API is not perfect, It could be possible some TOCTOU (I cannot think of any scenario right now), but it can be solved and while having a split functionality

Reply:
(Sorry, this is one of my comments that needs headlines to separate my rambling…)

Agreed with the comment on `.img`. (I think `.img` specifically was the reason why we never pursued using filename extensions in qemu.)

### Split probing functionality (format detection vs. opening)

Certainly imago could provide a `detect_format`, though asking that actually leads to another topic in itself.[1]

Providing such a split functionality in libkrun has the downside that libkrun won’t know whether the image format was probed or not, so the safeguards preventing malicious guests from formatting raw images need to either be on or off for all raw images all the time, regardless of whether the format was probed or not. (Of course, the user could enable/disable these safeguards with a flag, but then you can also just allow users to pass `NULL`/`None` as the format to make it probe.)

As you said, this would fully prevent guests e.g. from creating qcow2 images on raw block devices, if that’s what someone wants. (FWIW, may be a nested virtualization case. There are use cases of qcow2 on LVM.)

### To probe or not to probe

That said, I don’t want to sound like I’m in favor of probing. It’s proven itself to be a bad practice. But I do assume it’s a reality that must be lived with.

I’m afraid that if libkrun doesn’t provide a usable probing interface, people will find ways to do their own probing, which will be worse. Consider that imago doesn’t provide a probing interface, so this pull request implements it in libkrun.

I don’t think it’s unreasonable to decide to have libkrun not provide any probing, so it doesn’t get its hands dirty, while putting big fat warnings into its documentation, which may prevent its users from doing the wrong things, too. I just don’t have enough experience to know how effective such warnings really are in practice. (I should have done this for imago’s documentation, to see whether it works, as a social experiment. 😄)

### Probing in imago

[1] Putting any probing functionality into imago would beg some further questions:
- Returning an `enum` doesn’t work, because formats may be added, breaking compatibility. An opaque object with `.is_qcow2()` etc. would probably be best. (Could also be further extended to general image information.)
  - Honestly we’ll probably need some format representation for the “Is it OK to open this file?” callback anwyay.
- Why not allow imago to auto-open images in any format then? (Note: Already does this with backing files.)
- “If anyone does probing, it should be libkrun, because it controls the guest’s I/O path” applies to imago, too – why not move those safeguards into imago then? That would also save libkrun from duplicating a bit of probing code, specifically detecting the qcow2 magic sequence.

So really, same questions as we now have here for libkrun (“should it probe?”, “how to provide probing functionality?”, “should there be safeguards for probed raw images?”). I think any layer that were to implement probing would need to answer these questions.

Imago already has probing for qcow2 backing files, in case their format isn’t given in the qcow2 image. That’s not such a hot topic, though, because `qemu-img` these days requires users to put the backing format there; backing files are generally read-only; and so the only way to mess this up is basically by having the guest write e.g. a qcow2 image header to a raw format, and then taking a snapshot (so it becomes a backing file), without writing “raw” into the overlay’s backing format field (which I would consider quite wrong not to do).

Now, while the questions for how to probe are probably always the same regardless of on which layer you do it, maybe it’s easier to answer those questions if we consider probing were done in imago. Then, libkrun would be the user of this probing code, not the provider, so maybe it’s easier to know what you’d expect from it than to guess what users might expect from libkrun.

Reply:
> * Returning an `enum` doesn’t work, because formats may be added, breaking compatibility. An opaque object with `.is_qcow2()` etc. would probably be best. (Could also be further extended to general image information.)
Can't you use `#[non_exhaustive]` for this? This is what `ErrorKind` in the standard library uses.

> Why not allow imago to auto-open images in any format then? (Note: Already does this with backing files.)
At that point we could, but it would be good to have the callback you mentioned for implementing an allow/deny list of backing files. I think in the case of libkrun consumers it should be reasonable for them to have an explicit allow list (probably through a callback is fine, so this could even be used by an interactive application).

Reply:
> Can't you use #[non_exhaustive] for this?

True!

> At that point we could, but it would be good to have the callback you mentioned for implementing an allow/deny list of backing files.

Yes. I can work on it.

Reply:
> I don't think that is a good idea, it's like probing but worse. Also, extensions means nothing in a unix world, for instance, I have many qcow2 and raw images sharing the same extension `.img`

If that's standard practice to have qcow2 images with `.img` then yes my suggestion is a very bad idea.